### PR TITLE
ADD MODULE: gnu/sort

### LINKED ISSUE #3288: new module: GNU/SORT
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by DLBPointon**
## PR checklist

Closes #3288 

This PR adds a module for GNU/SORT. 
Test data has been added to the test-datasets repo and covers three test cases.

- [X] This comment contains a description of changes (with reason).
- [X] If you've fixed a bug or added code that should be tested, add tests!
- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] If necessary, include test data in your PR.
- [X] Remove all TODO statements.
- [X] Emit the `versions.yml` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [X] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/gnu/sort/main.nf [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    output:
+    tuple val(meta), file( "*.sorted" )   , emit: sorted
+    path "versions.yml"                    , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args        = task.ext.args     ?: ''
+    def output      = task.ext.prefix   ?: "${meta.id}.txt"
```
Comment (maxulysse) [inline modules/nf-core/gnu/sort/main.nf:22]: default extension is .txt? should it be all handled by the task.ext.prefix?

File: modules/nf-core/gnu/sort/main.nf [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    output_file     = "${output}.sorted"
```
Comment (maxulysse) [inline modules/nf-core/gnu/sort/main.nf:23]: I would tackle the `.sorted` part with the `task.ext.prefix`, or even use a `task.ext.suffix` for it.

File: modules/nf-core/gnu/sort/main.nf [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    def VERSION     = "9.1"             // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
+
+    """
+    cat ${file} | \\
+    sort ${args} > \\
+    ${output_file}
```
Comment (pontus) [inline modules/nf-core/gnu/sort/main.nf:29]:
```suggestion
    sort ${args} ${file} >  ${output_file}
```

I usually don't have a problem with superfluous cats, but here I think this looks nicer without it (and better on a single line).

File: modules/nf-core/gnu/sort/main.nf [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    END_VERSIONS
+    """
+
+    stub:
+    def args        = task.ext.args     ?: ''
+    def output      = task.ext.prefix   ?: "${meta.id}"
+    output_file     = "${output}.sorted"
+    def VERSION     = "9.1"
+    """
+    cat ${file} | sort ${args} > ${output_file}
```
Comment (pontus) [inline modules/nf-core/gnu/sort/main.nf:43]:
```suggestion
    sort ${args} ${file}> ${output_file}
```

File: modules/nf-core/gnu/sort/main.nf (hunk 3 above)
Comment (DLBPointon) [reply modules/nf-core/gnu/sort/main.nf:29]: You're right, it's how it needed to be used in the previous pipeline but no more. Just tested and works exactly the same.

File: modules/nf-core/gnu/sort/main.nf (hunk 2 above)
Comment (DLBPointon) [reply modules/nf-core/gnu/sort/main.nf:23]: Would this be a better implementation?
```
    output:
    tuple val(meta), file( "${output_file}" )   , emit: sorted
    path "versions.yml"                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args        = task.ext.args     ?: ''
    def prefix      = task.ext.prefix   ?: "${meta.id}"
    def suffix      = task.ext.prefix   ?: "${input.extension}"
    output_file     = "${prefix}.${suffix}"
    def VERSION     = "9.1"             // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.

    """
    sort ${args} ${input} > ${output_file}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        coreutils: $VERSION
    END_VERSIONS
    """

```

All works as expected locally

Comment (maxulysse) [reply modules/nf-core/gnu/sort/main.nf:23]: it should be this:
`suffix      = task.ext.suffix   ?: "${input.extension}"`

Comment (maxulysse) [reply modules/nf-core/gnu/sort/main.nf:23]: but yeah, that's perfect

Comment (DLBPointon) [reply modules/nf-core/gnu/sort/main.nf]: That's what I meant.

Nice, all sorted and committed now.

File: modules/nf-core/gnu/sort/main.nf (hunk 1 above)
Comment (DLBPointon) [reply modules/nf-core/gnu/sort/main.nf]: Referred to below.

**Comment by DLBPointon** [comment]
Thanks for the help @maxulysse  and @pontus !

**Pull request merged**